### ISSUE TITLE
Viewing Large Build Logs Takes A Very Long Time

### ISSUE THREAD

**Issue by astrieanna** [concourse/concourse#3521]
I tried to find a similar issue, since I've seen this on every team I've been on in the past two years, but I failed to find one.

## What challenge are you facing?

Our tests have very noisy logs (as any large ginkgo test suite run in verbose does). When we load build pages, it takes a significant time to load the text of the logs. In some cases, it can be very hard to tell if the page is "stuck" while loading or just taking forever to load. The worst is when you get frustrated and click on a link (to another build or back to the pipeline) because you think the text is just not loading, and then there's a long pause, and then the text appears, and then you immediately end up on the page you clicked the link to -- you were so close to your goal of reading that text, but it's gone and you have to start over on the "waiting to load text" timer if you go back to the build page.

## What would make this better?

It would help if the text displayed faster. It's just text, so I assume it's not a network/download problem, but a display slowness. The current concourse interface seems to just be slow to display a giant chunk of text, and I'm not familiar enough with how the UI is structured or where the time is spent to suggest a fix for that (other than that someone should be profiling the UI code).

## Are you interested in implementing this yourself?

Not at the moment.

**Comment by vito** [comment]
Yeah, this is a constant battle right now with our current architecture for the build logs. I feel this pain too, particular with builds that have pretty UIs like progress bars, which are cheap visually but costly at the transport level, as they work by constantly re-writing the same line in place.

The long and short of it is that ANSI escape sequences, the pseudo-protocol that determines how text renders in your terminal, are very hard to process "live" and render in a web browser. We've optimized the hell out of this in many different ways but there's only so much we can do without an architectural change. (Also, despite it being just text, sometimes it *is* a network/download speed problem, because of things like progress-bars and such.)

I'll be opening an issue soon to propose an architectural change that I think will help speed this up dramatically. For now I'll just throw this issue directly on the roadmap until I have time to write the other issue.

In the meantime, I have one bit of information to ask: is this a pain point you feel with finished builds, or also ones that are in-flight?

Thanks!

**Comment by astrieanna** [comment]
It's both, in that opening either can be very slow. The in-progress ones are worse because the currently running step is open by default, so you don't even get a choice about whether those are the logs you're even interested in. The loading is less bad on finished builds, but I think that's largely because you only pay the waiting cost when you expect it (i.e. when opening the tests task).

It's annoying on running builds when we're trying to poke at the environment to keep an eye on it during more speculative test runs. We look at the resource pulls and/or the first task to get information to target the environment, and this process is slowed down by the interface loading the huge log of the running tests and insisting on scrolling to the bottom of the log. (I mean, I see why that's a reasonable default -- opening builds to see what's currently going on is also common -- but its more annoying to pay that loading/scrolling cost when I just want to look at earlier, quick-to-load sections.)

**Comment by marco-m** [comment]
Related: #3499

**Comment by loghen41** [comment]
I was just about to create an issue similar to this, when I saw this one was already open.

@vito Is there an idea of timeline when this sort of enhancement would be put in? In relation to this, I know that previously when I was using Jenkins, it (somewhat) solved this issue through a sort of "tail" solution by only displaying the last 1000 lines of log text. It wasn't a smart fold, it simply just defaulted to a truncated log of the last 1000 lines.  A second click was required to load the entire file. 

The advantage to this approach is helpful in the situation like @astrieanna was facing when you are trying to just move quickly through different builds, and an accidental click makes you wait for an extremely long time.  That second click to expand the log file meant you were prepared to sit and wait for the entire file to load.

Is such a solution possible/advantageous for Concourse?

**Comment by vito** [comment]
@loghen41 Unfortunately not until we change how we record the data. We don't store by line, we store by event, and those events contain arbitrary chunks of data that have to be interpreted via an ANSI escape sequence processor before we'd even know the contents of the last 1000 lines. Right now that processing happens client-side via [`elm-ansi`](https://github.com/vito/elm-ansi), so a lot of time is spent on that, *and* a lot of time is spent just getting the raw events sent to the front-end in the first place.

@pivotal-jamie-klassen and I started chatting about a potential rearchitecture where we interpret those events server-side and store build logs in a much more sensible way. At that point I don't think we would even need windowing or tailing in order to render it efficiently, but it would also be much easier to implement that sort of thing if we need to. This would also make emitting the build logs to syslog drains a lot more natural.

**Comment by loghen41** [comment]
Ah, that makes more sense as to why the client-side almost completely locks up when trying to even process large text files.  That sounds like that solution would definitely handle that case much more efficiently.

**Comment by casparrolfe** [comment]
Would something like this help?
https://package.elm-lang.org/packages/FabienHenon/elm-infinite-list-view/latest/

**Comment by SimonXming** [comment]
We are hitting the same problem, the build log page completely stuck.
Chrome `Task Manager` indicate that particular `Tab` consume 2.7GB memory, it got paused because of `potential out-of-memory crash`.
I dump the whole build log into a `csv` file, total size is about 80MB.
We are really looking forwarded more efficient way of viewing log. :)

**Comment by evanchaoli** [comment]
@vito I would be "aggressively" pushing on this issue. I have noticed that many of pipelines on our clusters are generating huge build logs. As @SimonXming mentioned, some huge build logs even make Chrome browser crash. Fortunately FireFox is still able to load such build, so we have a workaround for now. But I think we should take priority to work on a solution. Otherwise if FireFox also crashes, we would be put into an awkward situation.

One thing we noticed is that Concourse UI uses too many DOM elements, while Jenkins uses a single DOM element for a log text box, which saves a lot of memories.

I have several ideas to optimize showing build logs:

1. Display build logs in a texbox rather than a table, so that uses a single DOM element to entire log block. (Can we have an option for task to consider that?)

2. By default, display only last 1000 lines of logs, meaning build log table will have at maximum 1000 rows. This wouldn't make log loading faster, but it would save browser's memory, so that at least prevents browsers from OOM. If user want to view whole logs, then open a new window to show while logs in plain text format.

3. Allow pipelines to define some log line patterns for tasks, then auto fold log lines that matches the patterns. I notices that a lot tasks generate duplicate logs like "Downloading file xxxx ...", a build may generate thousands lines of such trivial logs, which should be ok to be auto folded.

My team won't directly jump in to implement anything until some solutions/designs are agreed among us. But we are here!

**Comment by vito** [comment]
@evanchaoli We've prioritised this pretty highly in the UX backlog - it's the next epic after we do a quick implementation of build re-triggering (which shouldn't take long - MVP is just adding a button to the build page). We're itching to work on it. :slightly_smiling_face: 

Here's the roadmap epic for it: https://github.com/concourse/concourse/projects/37

**Comment by christophermancini** [comment]
This is an issue that comes up often, especially if we haven't deliberately optimized build output for a workflow.

**Comment by mlopez-eb** [comment]
Hi guys. I see the last message in this thread is from Oct 2020. Do we have any updates on this as my team is now running into the same issues. Log crashes Chrome, and Firefox has been spinning for about 10 minutes now.

This is a huge problem as our developers will not have access to containers for logs if they need them.

cc: @vito

**Comment by hfinucane** [comment]
A quick workaround you can use is to copy the URL and then drop it off to `fly`, like
```
fly watch -u https://my.concourse/...
```

You get full speed, full color logs in your terminal.

**Comment by dlbock** [comment]
Looks like this item was moved to the Icebox in May 2020? Does this mean it won't get addressed @vito ? 🙏

**Comment by dlbock** [comment]
> A quick workaround you can use is to copy the URL and then drop it off to `fly`, like
> 
> ```
> fly watch -u https://my.concourse/...
> ```
> 
> You get full speed, full color logs in your terminal.

@hfinucane For me, many times the output from `fly watch` is also somehow truncated

**Comment by taylorsilva** [comment]
This fell down the priority list, it's not on our roadmap to get done right now. [v10 stuff](https://blog.concourse-ci.org/core-roadmap-towards-v10/) has the highest priority for the team.

**Comment by wanderanimrod** [comment]
@taylorsilva , @vito mentioned having a conversation with someone about re-architecting the way logs are stored and processed. Is there a write-up for what they discussed somewhere? This is a huge pain for us at the moment. We'd like to invest some effort into remediating this. Any materials that would point us in the right direction will be highly appreciated. Thanks.

**Comment by taylorsilva** [comment]
I don't think that conversation happened. Did a quick search for issues and didn't find anything either. I personally haven't looked into the build logs much at all so I don't have anything I can really point you to. I'm also not super familiar with the frontend.

Maybe @aoldershaw has some thoughts or knows where some "quick" wins could be made.

**Comment by aoldershaw** [comment]
I'd take a look at https://github.com/concourse/concourse/issues/3879 which discusses some approaches. We will probably have to stop storing logs as build events and start interpreting them and storing them by line (and possibly not in Postgres, since it's really not the best place for it).

I wrote https://github.com/aoldershaw/ansi as a first step to interpret ANSI escape codes server-side - i ran some experiments just interpreting the stored build events before sending them to the client and saw some minor improvement, but nothing substantial in terms of rendering time. Using that library (or something like it) to interpet log messages and store them as formatted will allow us to e.g send the last 1000 log lines rather than everything which would improve performance a lot

**Comment by periardy** [comment]
Just adding a download link to the plain text version would do it. This is just insane to have this as an issue in 2022. Drop all the events content in a single file, at least we'll be able to look at the output of the build pipeline. While you discuss how to ANSI render text log on a web page. Please see BBS technology circa 1990.

**Comment by Anonymous-Coward** [comment]
There is, I believe, a simple solution, which only requires changes to the UI: don't display the long logs in their entirety inline, but put them in an iframe attached to the build step. This way, the iframe may be slow, but the pipeline's outline won't be. Maybe do some magic inside the iframe, which allows the logs to be loaded/displayed in chunks, with a minimal addition to the backend to allow querying for chunks by index. Alternatively, display just a tiny amount of logs, and have an ellipsis or something to display a textarea containing the entire log as plain text.

**Comment by Kump3r** [comment]
The issue still seems to be relevant in the latest Concourse versions. Is there a timeline, when this will be restructured, as to fix the problem? Thanks!

**Comment by taylorsilva** [comment]
No timeline on when this will be improved. If someone wants to make a PR that does something like:

- Detect if build logs are "big"
- Shows a link to download the raw logs or render the logs anyways

Then I'd happily review and accept it.

Reminder, that a workaround for viewing logs that kill your browser, you can use `fly watch -u https://ci.concourse/...` to view logs in your terminal

**Comment by jackjohnson-pm** [comment]
Not sure if others experience this, but for me even attempting to view the logs of large builds via `fly watch` in the terminal is super slow. Anyone else?

**Comment by zeroxer** [comment]
We are still hitting this on Concourse `8.2.1` (`worker_version: 2.5`) with a completed build log that is large but not absurdly large for CI output.

Observed behavior:

- Opening the finished build page in Chrome/Chromium causes the tab's JS heap to grow to ~3.1 GB and the page remains effectively unusable.
- The page is still mostly just the build outline at that point, with only ~129 DOM elements and no rendered log body, so this looks like client-side event/log parsing memory amplification rather than DOM size alone.
- Chrome later reports an error on the build event stream request.

Measurements from this build:

- `pipeline_build_events_<pipeline_id>` rows for the build: `5823`
- log events: `5812`
- logical payload size in Postgres: ~`166 MB` (`sum(length(payload))`)
- stored payload size in Postgres: ~`18 MB` (`sum(pg_column_size(payload))`)
- max individual payload: ~`35 KB`
- downloading `/api/v1/builds/<id>/events` produces ~`175 MB` of `text/event-stream`
- browser memory after opening the build page: `usedJSHeapSize ~= 3.15 GB`, `jsHeapSizeLimit = 4 GB`
- DOM at that point: ~`129` elements, `document.body.innerText.length ~= 245`

This was a C/C++/CPython build producing verbose compiler/linker output. The immediate workaround is to reduce stdout and use `fly watch`, but from a user perspective a single finished build page should not be able to crash the browser before the logs are even visibly rendered.

A few product-level mitigations that would make this much safer even before a full log storage rearchitecture:

- detect large build event streams and do not auto-open/render the log step inline;
- show a warning with options like "download raw log", "open last N lines", or "render anyway";
- provide a raw/plain-text log endpoint or download link that avoids client-side ANSI/event rendering;
- avoid fetching the full event stream until the user explicitly expands the large step.

There was also an nginx/openresty reverse proxy in front of this deployment, and proxy buffering/compression can make the transport behavior worse for SSE. But the core issue here is that the raw build event stream is ~175 MB and the Concourse UI can amplify that into multiple GB of browser heap.
